### PR TITLE
applications: nrf_desktop: Update connection latency handling on central

### PR DESCRIPTION
**Pull request by MarekPieta**
Change updates connection latency handling on nRF Desktop central.

### CODE REVIEW

File: applications/nrf_desktop/src/modules/ble_latency.c [hunk 1]
```diff
@@ -109,10 +114,14 @@ static void low_latency_check_fn(struct k_work *w)
 	}
 }
 
-static void le_param_updated(struct bt_conn *conn, u16_t interval,
-			     u16_t latency, u16_t timeout)
+static void conn_params_updated(struct ble_peer_conn_params_event *event)
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_latency.c:117]: `const struct ble_peer_conn_params_event *event`

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+		if (err == -EALREADY) {
+			/* Connection parameters are already set. */
+			err = 0;
+		}
+	}
+
+	return err;
+}
+
+static int update_peer_conn_params(struct connected_peer *peer)
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:107]: `const struct connected_peer *peer`

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+	if ((peer->llpm_support && (info.le.interval != CONN_INTERVAL_LLPM_REG)) ||
+	    (!peer->llpm_support && (info.le.interval != CONN_INTERVAL_BLE_REG)) ||
+	    (info.le.latency != peer->requested_latency)) {
+		return true;
+	}
+
+	return false;
+}
+
+static void conn_params_update_fn(struct k_work *work_desc)
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:157]: s/work_desc/work

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+{
+	__ASSERT_NO_MSG(work_desc != NULL);
+
+	int err = 0;
+
+	for (size_t i = 0; (!err && (i < ARRAY_SIZE(peers))); i++) {
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:163]: err in on peer should not affect others

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+		if (conn_params_update_required(&peers[i])) {
+			err = update_peer_conn_params(&peers[i]);
+		}
+	}
+}
+
+static void ble_peer_conn_params_event_handler(struct ble_peer_conn_params_event *event)
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:170]: const arg

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 4 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:163]: The only error reason that I observed was that controller was busy. In that case we simply need to retry after some time.
I may remove this to have more generic approach (if controller is busy we will only have one additional failed try).

File: applications/nrf_desktop/src/modules/ble_latency.c (hunk 1 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_latency.c:117]: done

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 2 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:107]: done

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 3 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:157]: done

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 5 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:170]: done

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+#define CONN_INTERVAL_LLPM_REG		0x0d01 /* 1 ms */
+#define CONN_INTERVAL_BLE_REG		0x0006 /* 7.5 ms */
+#define CONN_LATENCY_LOW		0
+#define CONN_LATENCY_DEFAULT		99
+#define CONN_SUPERVISION_TIMEOUT	400
+
+#define CONN_PARAMS_ERROR_TIMEOUT	K_MSEC(5)
+
+struct connected_peer {
+	struct bt_conn *conn;
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:32]: const

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+	bool llpm_support;
+	u16_t requested_latency;
+};
+
+static struct connected_peer peers[CONFIG_BT_MAX_CONN];
+static struct k_delayed_work conn_params_update;
+
+
+static struct connected_peer *find_connected_peer(struct bt_conn *conn)
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:41]: const arg

File: applications/nrf_desktop/src/modules/ble_conn_params.c [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+		if (!buf) {
+			LOG_ERR("Could not allocate command buffer");
+			return -ENOBUFS;
+		}
+
+		u16_t conn_handle;
+
+		err = bt_hci_get_conn_handle(conn, &conn_handle);
+		if (err) {
+			LOG_ERR("Failed obtaining conn_handle (err:%d)", err);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_conn_params.c:74]: `buf` is not released on error

Comment (pdunaj) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:74]: `net_buf_unref` should be called on it most probably

Comment (pdunaj) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:74]: or `bt_hci_get_conn_handle` should be called before the buffer is obtained

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 6 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:32]: done

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 7 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:41]: done

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 6 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:32]: `bt_conn_le_param_update` takes `struct bt_conn *conn` as an argument.
The connection object is modified there, so we should not use the const pointer in the module.

File: applications/nrf_desktop/src/modules/ble_conn_params.c (hunk 8 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/ble_conn_params.c:74]: You are right. I reordered the calls.
It was leftover from previous implementation.

**Pull request merged**